Android drawable perf

albyrock87 opened the pull request:
### Description of Change

Backgrounds are everywhere in our apps, so we must be fast when drawing android views, and `Java -> C#` is not.

This PR tries to do its best to completely move the logic into a java `PlatformDrawable` class.


### Benchmarks (~50% improvements in mapping, ~60% improvement in rendering time)

I used #24414 as main page wrapped into a `Shell` component given it has a `drawChild` method which shows the cumulative draw time.

The results are impressive and will for sure make an impact!

[24414-net10.json](https://github.com/user-attachments/files/22296183/24414-net10.json)
<img width="754" height="144" alt="image" src="https://github.com/user-attachments/assets/834b59e0-77f9-457c-9867-179b91f116d0" />
<img width="749" height="46" alt="image" src="https://github.com/user-attachments/assets/5ecf5252-10c8-42e8-85f0-b94dfd60746b" />
<img width="752" height="124" alt="image" src="https://github.com/user-attachments/assets/04d78a16-f22d-4ff1-ab5f-b34316c7240a" />

[24414-pr-final2.json](https://github.com/user-attachments/files/22403331/24414-pr-final2.json)
<img width="1106" height="122" alt="image" src="https://github.com/user-attachments/assets/bc59bef4-9e2f-4626-b399-7ab0b3b44cd0" />
<img width="918" height="104" alt="image" src="https://github.com/user-attachments/assets/caca6cba-69c4-4949-b7df-cac5ceccb6d9" />
<img width="903" height="123" alt="image" src="https://github.com/user-attachments/assets/2b6e9b52-c347-43d8-abec-061263739724" />

albyrock87 committed:
Improve Android background rendering performance

jsuarezruiz reviewed (changes requested):
There are some related failing tests:
<img width="338" height="234" alt="image" src="https://github.com/user-attachments/assets/a1b5e5ec-fb5a-436d-b31e-209c5f571b99" />

Differences between snapshots. Based on the failing tests, seems like not apply dynamic changes (like a dynamic binding for example. Could you review it? Let me know if can help with something.

albyrock87 committed:
Fix gradient not working

albyrock87 committed:
Fix undetected shape changes

albyrock87 committed:
Update screenshots

albyrock87 commented:
@jsuarezruiz @PureWeen the PR is now ready, I added a few commits with the fixes and updated screenshots, here's a summary:
- `IShape` can change even without changing instance (because `Shape` is a bindable property`) so for performance reasons, I've introduced an internal `IVersionedShape` so that the `MauiDrawable` can still avoid some calls and retain the expected performance
- Screenshots are different because I finally fixed the bug where the default stroke color was black on Android (you can verify other platforms and you will see the border is transparent unless the developer specifies a color)
- Fixed gradients not working

mattleibow commented:
/azp run

albyrock87 committed:
Fixes remaining tests

albyrock87 committed:
Add missing public API on Windows

albyrock87 commented:
Do not merge. I'm working on feedbacks.

albyrock87 committed:
Improve performance by avoiding arrays as much as possible

albyrock87 commented:
The PR is now ready!

albyrock87 committed:
Remove old device tests

albyrock87 commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformDrawable.java:
```diff
@@ -0,0 +1,284 @@
+
+    // Simplified style management
+    public void setLinearGradientBackground(float x1, float y1, float x2, float y2, int[] colors, float[] positions) {
+        float[] bounds = { x1, y1, x2, y2 };
```
Such validation wasn't there since the beginning, and I feel this would be the wrong place to add it given we're already in the Java world and we need this to be as fast as possible.

A better place would be at .NET level so that every platform could benefit from this check, though I'm not sure what would be the best place to do it.

albyrock87 committed:
Improve documentation

PureWeen commented:
/azp run

bcaceiro commented:
Any news on this?

albyrock87 committed:
Merge branch 'main' into android-drawable-perf

# Conflicts:
#	docs/design/UITesting.md
#	src/Controls/src/Core/PublicAPI/net-android/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-ios/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-tizen/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-windows/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/netstandard/PublicAPI.Unshipped.txt
#	src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt

albyrock87 committed:
Merge branch 'main' into android-drawable-perf

# Conflicts:
#	src/Controls/src/Core/PublicAPI/net-ios/PublicAPI.Unshipped.txt
#	src/Controls/src/Core/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt

albyrock87 commented:
I won't keep solving conflicts every time, so please let me know when you really plan to merge this so I can do the work just once. Thanks :)

PureWeen committed:
Merge remote-tracking branch 'origin/main' into android-drawable-perf

# Conflicts:
#	src/Controls/src/Core/Shapes/Shape.cs
#	src/Core/src/PublicAPI/net-android/PublicAPI.Unshipped.txt

PureWeen commented on src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformDrawableStyle.java:
```diff
@@ -0,0 +1,165 @@
+            this.hasNoneColor = 1;
+            this.noneColor = value.data;
+        }
+
```
should there be an else here?

albyrock87 replied:
Good catch, I've added a `return;` above.

albyrock87 committed:
Update src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformDrawableStyle.java

albyrock87 committed:
Fix issue with theme color

albyrock87 commented:
So it turns out that on `main` there was always a big mistake

https://github.com/dotnet/maui/blob/e6bb31a21765d091048c79ea128ead97ef877a07/src/Core/AndroidNative/maui/src/main/java/com/microsoft/maui/PlatformInterop.java#L510-L518

As you can see there is a `!` in front of `context.getTheme().resolveAttribute` which is obviously wrong.
In my code this was correct, but I had another [bug which I fixed in the last commit](https://github.com/dotnet/maui/pull/31567/changes/f9af3fde469863e9ec9259390ec9b373e092548f).

Correcting my bug though created a different behavior because the code on `main` has never worked 🐛 

Given this never worked, I have completely removed the theme-related code which makes everything work fine and the code cleaner.

You can see the changes in my [last commit](https://github.com/dotnet/maui/pull/31567/commits/3582b397e601601026fc3b38d47bc6ddc6215b97).

jfversluis committed:
Merge branch 'inflight/current' into android-drawable-perf
